lib/jsonapi/utils/string_manipulation.ex

```diff
@@ -0,0 +1,148 @@
+  @moduledoc """
+  String manipulation helpers.
+  """
+
+  @doc """
+  Replace dashes between words in `value` with underscores
+
+  Ignores dashes that are not between letters/numbers
+
+  ## Examples
+
+      iex> underscore("top-posts")
+      "top_posts"
+
+      iex> underscore("-top-posts")
+      "-top_posts"
+
+      iex> underscore("-top--posts-")
+      "-top--posts-"
+
+      iex> underscore(%{"foo-bar" => "baz"})
+      %{"foo_bar" => "baz"}
+
+      iex> underscore({"foo-bar", "dollar-sol"})
+      {"foo_bar", "dollar-sol"}
+
+      iex> underscore({"foo-bar", %{"a-d" => "z-8"}})
+      {"foo_bar", %{"a_d" => "z-8"}}
+
+      iex> underscore(%{"f-b" => %{"a-d" => "z"}, "c-d" => "e"})
+      %{"f_b" => %{"a_d" => "z"}, "c_d" => "e"}
+
+      iex> underscore(:"foo-bar")
+      :foo_bar
+
+      iex> underscore(%{"f-b" => "a-d"})
+      %{"f_b" => "a-d"}
+  """
+  def underscore(value) when is_binary(value) do
+    String.replace(value, ~r/([a-zA-Z0-9])-([a-zA-Z0-9])/, "\\1_\\2")
+  end
+
+  def underscore(map) when is_map(map) do
+    Enum.into(map, %{}, &underscore/1)
+  end
+
+  def underscore({key, value}) when is_map(value) do
+    {underscore(key), underscore(value)}
+  end
+
+  def underscore({key, value}) do
+    {underscore(key), value}
+  end
+
+  def underscore(value) when is_atom(value) do
+    value
+    |> to_string()
+    |> underscore()
+    |> String.to_atom()
+  end
+
+  def underscore(value) do
+    value
+  end
+
+  @doc """
+  Replace underscores between words in `value` with dashes
+
+  Ignores underscores that are not between letters/numbers
+
+  ## Examples
+
+      iex> dasherize("top_posts")
+      "top-posts"
+
+      iex> dasherize("_top_posts")
+      "_top-posts"
+
+      iex> dasherize("_top__posts_")
+      "_top__posts_"
+  """
+  def dasherize(value) when is_atom(value) do
+    value
+    |> to_string()
+    |> dasherize()
+  end
+
+  def dasherize(value) when is_binary(value) do
+    String.replace(value, ~r/([a-zA-Z0-9])_([a-zA-Z0-9])/, "\\1-\\2")
+  end
+
+  def dasherize(%{__struct__: _} = value) when is_map(value) do
+    value
+  end
+
+  def dasherize(value) when is_map(value) do
+    Enum.into(value, %{}, &dasherize/1)
+  end
+
+  def dasherize({key, value}) do
+    if is_map(value) do
+      {dasherize(key), dasherize(value)}
+    else
+      {dasherize(key), value}
+    end
+  end
+
+  @doc """
+  Whether or not outgoing JSON should be dashed.
+
+  Configure JSONAPI to dasherize outgoing fields by setting `:underscore_to_dash`
+  to true.
+
+  ## Example
+
+      # Dasherize outgoing fields
+      config :jsonapi,
+        underscore_to_dash: true
+
+      # Outgoing fields should be underscored, this is the default
+      config :jsonapi,
+        underscore_to_dash: false
+  """
+  def dasherize?, do: Application.get_env(:jsonapi, :underscore_to_dash, false)
+
+  @doc """
+  Dasherizes a String if, and only if, `dasherize?/0` is true
+  """
+  def maybe_dasherize(term) do
```

Comment:
I'm a bit undecided on rolling these methods in. They feel awfully private to me, or like they shouldn't exist at all. I may pull them back into their old modules and mark them as private.

Reply:
I would prefer we drop prefixes like `maybe_` at the very least.

Reply:
Well, one option we have is to keep at least one of these for now. If we introduce a Plug that transforms OUTGOING params we needn't any of this at all. This would also mean we could have outgoing/incoming pairs for things like having a dasherized or camelized API.